Synrgy 134

LHCPlayz opened the pull request:
Feature: Implement Fetch User Details API

Adds method to fetch user details from id stored in JWT token.

LHCPlayz committed:
SYNRGY-134 - Initial attempt

Still figuring out how to set up the database connection with Prisma and PostgreSQL, so I haven't tested this with the database. Also I will include all additional requirements for this feature as soon as I properly figured out most of the tools.

LHCPlayz committed:
SYNRGY-134 - Update

Tried to test the function; but that doesn't seem to be working at this point in time. Will try again tomorrow.

LHCPlayz committed:
SYNRGY-134 - Update

I have gotten the feature to work at this stage. Now I will be implementing the extra bits.

LHCPlayz committed:
SYNRGY-134 - 3/4 finished

Currently everything is here except endpoint verification. I will speak to Pablo and ask him about it.

LHCPlayz committed:
SYNRGY-134 - Comments and tweaks

Commented code and some minor tweaks

LHCPlayz committed:
SYNRGY-134 - JWT migration in progress

Adding the JWT authentication to getuserinfo. Still working on that.

LHCPlayz committed:
SYNRGY-134 - Prisma database schema update + authentication testing.

Not quite done yet. Still have lots of errors.

LHCPlayz committed:
SYNRGY-134 - All features present

All features are present.

When testing the user recall function uncomment the test JWT token and ensure you have data in your database under table User with id = 1

j9mmy reviewed (changes requested):
Make use of Prisma ORM instead of writing raw SQL queries and manually establishing a connection to the database. Would downsize the current code by a ton.

Also, but could be dismissed, converting the "userCheckerRouter" to "userRouter" would make more sense as an entire route dedicated to checking a user would be unnecessary. I'd advise to rename the route in the `app.js` to be `/api/users` as well so you don't constantly have to state `/users/...` in each function of the user route file.

LHCPlayz committed:
SYNRGY-134 - Changes

- Using Prisma instead of raw Sql

- Converted "userCheckerRouter" to "userRouter"

-Renamed the route in app.js to /api/users

- Changed the primsa.schema file to match current database format.

- Changed the migration.sql file to match the current databse format according to documentation.

LHCPlayz committed:
Merge branch 'develop' into SYNRGY-134

LHCPlayz commented:
- Using Prisma instead of raw Sql

- Converted "userCheckerRouter" to "userRouter"

- Renamed the route in app.js to /api/users

- Changed the primsa.schema file to match current database format.

- Changed the migration.sql file to match the current databse format according to documentation.

Remember to run "npx prisma generate" and "npx prisma migrate dev --name init" before testing.

j9mmy commented on src/app.js:
```diff
@@ -1,8 +1,10 @@
 const app = express();
 app.use(express.json());
 app.use('/api', healthCheckerRouter);
+app.use('/api', userRouter);
```
> Renamed the route in app.js to /api/users

...You sure?

LHCPlayz replied:
I'd blame it on a lack of coffee; but it's just pure skill issue. 
I'm sure now lol... Hopefully...

j9mmy replied:
💀

LHCPlayz committed:
Update app.js

j9mmy reviewed (changes requested):
I suggest splitting up the `user.js` into controller and service layers. Pablo has a good minimal example in #9. In short: controllers handle the response statuses, and services are the actual logic.

To add to the point above, notice the folder structure of the example. Currently you made a `classes` folder in the root-level of the project, but that is not optimal. Try and put it within the `src` folder instead. In the aforementioned example there is a `controllers` and `services` folder within the `src` folder, so I suggest making those to replace the classes folder.

Just a thing to note. If you correctly follow the previous example, then you can get rid of the constructor. Even less code, so there's that.

j9mmy reviewed (commented):
Forgot to mention, but keep the code within the `routes` folder minimal. In the current code you have Controller logic inside of the `routes` folder, but it would be more optimal to transfer this code to it's own respective `controllers` folder. So the only thing you would need to do in `routes/user.js` is reference the function, with the controller logic you currently have, within the `controllers` folder.

LHCPlayz committed:
SYNRGY-134 - Service and controller layers

- Removed classes folder
- split user.js into controller and service layers
- removed the controller logic inside of the routes/user.js file

j9mmy reviewed (approved):
Depending on who uses this feature, this approval may not be valid. Currently you are returning the entire user object, so that includes **all** columns of the user table. 

- If a user wants to ask their **own** user details then it'd be fine, albeit with redundant data like the ID of the user (pretty sure the user doesn't care about that).
- If an admin wants to see the details of a certain user, then retrieving all of the columns of the user table makes sense.
- If a user wants to snoop on other user's details, then it'd be problematic since, other than redundant data, it also returns sensitive data like the password (albeit hashed).

I am not sure how specific this implementation needs to be, but if the above "problems" are no issue, then I see it as approved.

_In case you don't know... If you want to only partially return the user columns to exclude redundant/sensitive data, then you can simply create a new object with the columns you'd only like to pass on like so:_
```
const filteredUser = {
  name = user.name,
  email = user.email,
  ...
}
```

_Simply pass `filteredUser` instead of `user` in this case._
